TouchPortalAPI/TpToPy.py

```diff
@@ -0,0 +1,140 @@
+        return generatedAction
+
+    def generateEvents(self):
+        generatedEvents = {}
+        categories = self.entry.get("categories", [])
+
+        for category in range(len(categories)):
+            if categories[category].get("events", False) and isinstance(categories[category]['events'], list):
+                for event in range(len(categories[category]["events"])):
+                    generatedEvents[event+1] = categories[category]["events"][event]
+                    generatedEvents[event+1]["category"] = categories[category].get("id", "").split(".")[-1]
+
+        return generatedEvents
+
+    def generateConnectors(self):
+        generatedConnectors = {}
+        categories = self.entry.get("categories", [])
+        for category in range(len(categories)):
+            if categories[category].get("connectors") and isinstance(categories[category]['connectors'], list):
+                for connector in range(len(categories[category]["connectors"])):
+                    generatedConnectors[connector+1] = categories[category]["connectors"][connector]
+                    generatedConnectors[connector+1]["category"] = categories[category].get("id", "").split(".")[-1]
+                    if generatedConnectors[connector+1].get('data', False):
+                        generatedConnectors[connector+1]['format'] = self.__convertFormat(generatedConnectors[connector+1]['format'], generatedConnectors[connector+1]['data'])
+                        generatedConnectors[connector+1]['data'] = self.__convertData(generatedConnectors[connector+1]['data'])
+        
+        return generatedConnectors
+    
+    def generateCalegories(self):
+        generatedCategories = {}
+        categories = self.entry.get("categories", [])
+
+        for category in categories:
+            categId = category.get("id", "").split(".")
+            generatedCategories[categId[-1]] = {
+                "id": ".".join(categId[:-1]) or categId[-1],
+                "name": category.get("name", ""),
+                "imagepath": category.get("imagepath", ""),
+            }
+        return generatedCategories
+
+    def writetoFile(self, fileName):
+        TP_PLUGIN_INFO = self.generateInfo()
+        TP_PLUGIN_SETTINGS = self.generateSettings()
+        TP_PLUGIN_STATES = self.generateStates()
+        TP_PLUGIN_ACTIONS = self.generateActions()
+        TP_PLUGIN_CONNECTORS = self.generateConnectors()
+        TP_PLUGIN_EVENTS = self.generateEvents()
+        TP_PLUGIN_CATEGORIES = self.generateCalegories()
+        with open(fileName, 'w') as f:
+            f.write("#!/usr/bin/env python3\n")
+            for entryVar in ["TP_PLUGIN_INFO", "TP_PLUGIN_SETTINGS", "TP_PLUGIN_CATEGORIES", "TP_PLUGIN_CONNECTORS", "TP_PLUGIN_ACTIONS", "TP_PLUGIN_STATES", "TP_PLUGIN_EVENTS"]:
+                struct = json.dumps(locals()[entryVar], indent=4, sort_keys=False, skipkeys=True)\
```

Comment:
```py
struct = json.dumps(....)
struct = re.sub(r'": (true|false)(,?)\n', lambda m: f'": {m.group(1).title()}{m.group(2)}\n', struct)
```
Advantage is it only scans the generated text once.

Reply:
Added